tests/test_cargo_version_range.py

```diff
@@ -0,0 +1,127 @@
+import pytest
+
+from univers.version_range import CargoVersionRange
+from univers.version_range import InvalidVersionRange
+from univers.versions import CargoVersion
+
+values = [
+    # https://doc.rust-lang.org/cargo/reference/specifying-dependencies.html
+    # caret
+    ["^1.2.3", [[]], ["1.2.3"], []],
+    # tilde
+    ["~1.2.3", [[">=", "1.2.3"], ["<", "1.3.0"]], ["1.2.4"], ["2.0.1"]],
+    ["~1.2", [[">=", "1.2.0"], ["<", "1.3.0"]], ["1.2.5"], ["1.3.1"]],
+    [
+        "~1",
+        [[">=", "1.0.0"], ["<", "2.0.0"]],
+        ["1.3.0", "1.8.1"],
+        ["2.1.0", "2.2"],
+    ],  # tilde increment the major
+    # wildcard
+    ["*", [[">=", "0.0.0"]], ["1.0.0", "2.0.0"], []],
+    ["1.*", [], ["1.0.0"], ["2"]],
+    ["1.2.*", [[">=", "1.2.0"], ["<", "1.3.0"]], ["1.2", "1.2.1"], ["2.1.0", "2.2"]],
+    # https://github.com/dtolnay/semver/blob/master/tests/test_version_req.rs :
+    # test_basic
+    # ["^1.0.0", [[]], ["1.0.0", "1.1.0", "1.0.1"], ["0.9.9", "0.10.0", "0.1.0", "1.0.0-pre", "1.0.1-pre"]],
+    # test_exact
+    ["=1.0.0", [["=", "1.0.0"]], ["1.0.0"], ["1.0.1", "0.9.9", "0.10.0", "0.1.0", "1.0.0-pre"]],
+    ["=0.9.0", [["=", "0.9.0"]], ["0.9.0"], ["0.9.1", "1.9.0", "0.0.9", "0.9.0-pre"]],
+    ["=0.0.2", [["=", "0.0.2"]], ["0.0.2"], ["0.0.1", "0.0.3", "0.0.2-pre"]],
+    [
+        "=0.1.0-beta2.a",
+        [["=", "0.1.0-beta2.a"]],
+        ["0.1.0-beta2.a"],
+        ["0.9.1", "0.1.0", "0.1.1-beta2.a", "0.1.0-beta2"],
+    ],
+    # https://github.com/dtolnay/semver/blob/master/tests/test_version_req.rs#L73
+    # ["=0.1.0+meta", [["=", "0.1.0+meta"]], ["0.1.0", "0.1.0+meta", "0.1.0+any"], []],
+    # test_greater_than
+    # [">= 1.0.0", [[]], ["1.0.0", "2.0.0"], ["0.1.0", "0.0.1", "1.0.0-pre", "2.0.0-pre"]],
+    # [">= 2.1.0-alpha2", [[]], ["2.1.0-alpha2", "2.1.0-alpha3", "2.1.0", "3.0.0"], ["2.0.0", "2.1.0-alpha1", "2.0.0-alpha2", "3.0.0-alpha2"]],
+    # test_less_than
+    # ["<1.0.0", [[]], ["0.1.0", "0.0.1"], ["1.0.0", "1.0.0-beta", "1.0.1", "0.9.9-alpha"]],
+    # ["<= 2.1.0-alpha2", [[]], ["2.1.0-alpha2", "2.1.0-alpha1", "2.0.0", "1.0.0"], ["2.1.0", "2.2.0-alpha1", "2.0.0-alpha2", "1.0.0-alpha2"]],
+    # [">1.0.0-alpha, <1.0.0", [[">", "2.1.0-alpha2"], ["<", "1.0.0"]], ["1.0.0-beta"], []],
+    # [">1.0.0-alpha, <1.0", [[">", "1.0.0-alpha"], ["<", "1.0"]]], ["1.0.0-beta"], []],
```

Comment:
@ziadhany why're we not testing these Cargo version ranges?

Reply:
I commented `# ` all the tests that generate errors , so I can try to bypass the test one by one .

Reply:
@ziadhany It appears like we're having issues with the pre-release versions. The Cargo Semver library (https://github.com/dtolnay/semver) is primarily meant for processing version range in the Cargo manifest. Consequently, any version range will not include the pre-release version unless one of the range boundaries is a pre-release (e.g., ">1.0.0-alpha, <1.0"). See https://doc.rust-lang.org/cargo/reference/resolver.html#pre-releases for more details.

However, for our use case in VulnerableCode, we do want the pre-release to be included in the version range. We have encountered similar behavior with NPM ranges, and it's not surprising that the Cargo version range is mostly based on NPM (https://github.com/dtolnay/semver/issues/58).

Let's discuss this in our upcoming community call.

Reply:
yes the main problem with Cargo version ranges is the pre-release versions

[rustsec/advisory-db](https://github.com/rustsec/advisory-db) use the pre-release version range :
https://github.com/rustsec/advisory-db/blob/a5fb72de318a74eb69a2c241c0e46705684a35d0/crates/lettre/RUSTSEC-2021-0069.md?plain=1#L12

so I think we should have a way to parse the pre-release versions in Univers